src/server-glue/webresource-handler.ts

```diff
@@ -0,0 +1,355 @@
+import type * as Express from 'express';
+import * as busboy from 'busboy';
+import * as is from 'type-is';
+import * as stream from 'stream';
+import * as uriParser from '../sbvr-api/uri-parser';
+import { getApiRoot, getModel } from '../sbvr-api/sbvr-utils';
+import { checkPermissions } from '../sbvr-api/permissions';
+import * as sbvrUtils from '../sbvr-api/sbvr-utils';
+import { S3Handler } from './webresource-handlers/S3Handler';
+import { NoopHandler } from './webresource-handlers/NoopHandler';
+import {
+	odataNameToSqlName,
+	sqlNameToODataName,
+} from '@balena/odata-to-abstract-sql';
+import { errors, permissions } from './module';
+
+export interface IncomingFile {
+	fieldname: string;
+	originalname: string;
+	encoding: string;
+	mimetype: string;
+	stream: stream.Readable;
+}
+
+export interface UploadResponse {
+	size: number;
+	filename: string;
+}
+
+export interface WebResourceHandler {
+	handleFile: (resource: IncomingFile) => Promise<UploadResponse>;
+	removeFile: (fileReference: string) => Promise<void>;
+}
+
+export class WebResourceError extends Error {
+	name = 'WebResourceError';
+	constructor(message: string, originalError?: Error) {
+		super(message);
+
+		if (originalError) {
+			this.originalError = originalError;
+		}
+	}
+
+	originalError?: Error;
+}
+
+export class FileSizeExceededError extends WebResourceError {
+	name = 'FileSizeExceededError';
+	constructor(maxSize: number) {
+		super(`File size exceeded the limit of ${maxSize} bytes.`);
+	}
+}
+
+type WebResourcesDbResponse = {
+	[fieldname: string]: { href: string } | undefined | null;
+};
+
+const ifFileInValidPath = async (
+	fieldname: string,
+	req: Express.Request,
+): Promise<boolean> => {
+	if (req.method !== 'POST' && req.method !== 'PATCH') {
+		return false;
+	}
+
+	const apiRoot = getApiRoot(req);
+	if (apiRoot == null) {
+		return false;
+	}
+	const model = getModel(apiRoot);
+	const { resourceName } = await uriParser.parseOData({
+		url: req.url,
+		method: req.method,
+	});
+
+	const permission = req.method === 'POST' ? 'create' : 'update';
+	const vocab = model.versions[model.versions.length - 1];
+	const hasPermissions = await checkPermissions(
+		req,
+		permission,
+		resourceName,
+		vocab,
+	);
+
+	if (!hasPermissions) {
+		return false;
+	}
+
+	// TODO: This could be cached
```

Comment:
@fisehara Actually no. All the todos left were things I thought while coding this "If we run into performance issues, this is where it could be improved". My understanding is that a lot of caching is already going on in the background and adding more on top of it just for the sake of "it might be a performance improvement" would make it even harder to mantain. But I also think that lots of these TODOs around look ugly/incomplete. What do you think of removing them from the code and storing as tips somewhere in case we need to revisit/improve performance?